WIP: next step for documentation of groups

ThomasBreuer opened the pull request:
- changed some docstrings
- changed the code where it is necessary for documenting it
  (e.g., admit a type as optional first argument for `exponent`, `index`)
- export `ischaracteristic` not `ischaracteristic_subgroup`
  (the latter is not needed)
- extended and adjusted tests

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -396,9 +505,9 @@ end
+    ngens(G::GAPGroup) -> Int
 
-Return the length of the array gens(G).
+Return the length of the array `gens(G)`, see [`gens`](@ref).
```
Would this work?
```suggestion
Return the length of the array [`gens`](@ref)`(G)`.
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -840,12 +947,14 @@ end
-Return the nilpotency class of `G`, that is the smallest integer `d` such that `G` has a central series of length `n`.
+Return the nilpotency class of `G`, that is, the smallest integer `d` such that `G` has a central series of length `n`.
+
+An error is signalled if `G` is not nilpotent.
```
```suggestion
An exception is thrown if `G` is not nilpotent.
```

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -139,23 +161,28 @@ end
-- `free_group(L::String...) -> FPGroup`; return the free group with length(`L`) generators, printed as `L[1]`, `L[2]`, `L[3]`, etc.
-- `free_group(L::Array{String,1}) -> FPGroup`; same as above.
-- `free_group(n::Int, s::String) -> FPGroup`; return the free group of rank `n`, with generators printed as `"s1"`, `"s2"`, `"s3"`, etc.
+The second form, if there are `n` arguments `L...`,
```
Maybe this was meant?
```suggestion
The third form, if there are `n` arguments `L...`,
```

ThomasBreuer replied:
Yes.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -80,32 +81,69 @@ function parent(x::GAPGroupElem)
 
-order(x::Union{GAPGroupElem, GAPGroup}) = order(fmpz, x)
+"""
+    order([::Type{T}, ]x::Union{GAPGroupElem, GAPGroup}) where T
```
This is inspired by GAPDoc convention, where we signal that the first argument is optional. Question is whether this is clear for other people. @fieker @thofma and others: thoughts?

An alternative would be to imiate the Julia convention for default values:
```suggestion
    order(::Type{T} = fmpz, x::Union{GAPGroupElem, GAPGroup}) where T
```

Third alternative would be to list both signatures:
```suggestion
    order(::Type{T}, x::Union{GAPGroupElem, GAPGroup}) where T
    order(x::Union{GAPGroupElem, GAPGroup})
```

thofma replied:
According to https://docs.julialang.org/en/v1/manual/documentation/ (first item), `[ ... ]` is used for optional arguments without default value?

ThomasBreuer replied:
Yes. Thus the first of the two above alternatives fits (and also has the advantage that the default need not be explained in the text).

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -80,32 +81,69 @@ function parent(x::GAPGroupElem)
 
-order(x::Union{GAPGroupElem, GAPGroup}) = order(fmpz, x)
+"""
+    order([::Type{T}, ]x::Union{GAPGroupElem, GAPGroup}) where T
```
Orthogonal to all this: should we restrict `T` to be `Union{Integer, fmpz}` ? And: do we already have a type alias for this probably very common union?

fingolfin commented on src/Groups/libraries/smallgroups.jl:
```diff
@@ -70,7 +70,15 @@ function all_small_groups(n::Int, L...)
+#T all_small_groups( 60 )  -> Array{Oscar.GAPGroup,1}
+#T all_small_groups( 59 )  -> Array{PcGroup,1}
 
+#T Do we want this???
```
I will partially deal with this in my PR for the group libraries, by turning this into an iterator.

ThomasBreuer replied:
Thanks.

fingolfin commented:
@ThomasBreuer What is the status of this? I am somewhat surprised to see that it's still not complete after almost two weeks? What's the holdup?

(Also, now it needs to be rebased...)

ThomasBreuer committed:
next step towards documentation for groups

- fixed a few `@docs` blocks in `docs` files such that the correct docstrings are specified
- moved text from `docs` files to `src` files where appropriate
- made statements about exceptions uniform (say "An exception is thrown if ...")
- used `@gapattribute` in more cases
- added/extended/corrected docstrings
- changed the syntax in lines that show the signatures of the functions such that default values are shown, and removed the related (now unnecessary) sentences from the text.
- use LaTeX markup for math mode where appropriate

ThomasBreuer commented:
@fingolfin Now it is ready for review. Sorry for the delay.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -85,68 +86,191 @@ function parent(x::GAPGroupElem)
 
-Return a random element of the group `G`. It works faster than `rand`, but the elements are not necessarily equally distributed.
+Return a pseudo random element of `G`.  This works faster than `rand`,
+but the returned elements are not necessarily equally distributed.
```
```suggestion
but the returned elements are not necessarily uniformly distributed.
```

ThomasBreuer replied:
Yes.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -178,41 +302,33 @@ end
-function ==(x::PermGroupElem, y::PermGroupElem)
-   return x.X == y.X && degree(parent(x))==degree(parent(y))
-end
+==(x::PermGroupElem, y::PermGroupElem) = x.X == y.X && degree(parent(x))==degree(parent(y))
```
I realize you are just refactoring the existing code, but while we are at it: perhaps we should first check the degree (as it ought to be fast to get) before comparing the permutations? I.e.,
```suggestion
==(x::PermGroupElem, y::PermGroupElem) = degree(parent(x))==degree(parent(y)) && x.X == y.X
```

ThomasBreuer replied:
Yes.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -343,13 +482,27 @@ julia> cperm([1,2,3],4:7)
+At the moment, the input vectors of the function `cperm` need not be disjoint.
+
+!!! warning
+    If the function `perm` is evaluated in a vector of integers without specifying the group `G`, then the returned value is an element of the AbstractAlgebra.jl type `Perm{Int}`. For this reason, if one wants a permutation of type `GAPGroupElem{PermGroup}` without specifying a parent, one has to use the function `gap_perm`.
```
Very long line, maybe wrap it?

ThomasBreuer replied:
Yes.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -748,9 +947,10 @@ function sylow_subgroup(G::GAPGroup, p::Int64)
+    hall_subgroup(G::Group, P::Vector{Int64})
 
-Return a Hall `P`-subgroup of `G`. It works only if `G` is solvable.
+Return a Hall `P`-subgroup of `G`.
```
A definition of "Hall subgroup" would be nice, possibly taken from the GAP manual. Likewise for an example (again we could borrow from the GAP manual).

ThomasBreuer replied:
Yes.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -748,9 +947,10 @@ function sylow_subgroup(G::GAPGroup, p::Int64)
 
-Return a Hall `P`-subgroup of `G`. It works only if `G` is solvable.
+Return a Hall `P`-subgroup of `G`.
+An exception is thrown if `G` is not solvable.
```
This indeed describes what the code does, but... why actually does it? In GAP, I think we support non-solvable groups, too, at least if they have suitable subgroups? Perhaps we don't want to support this here -- fine, but then we should perhaps still elaborate *why*, i.e. why we throw an exception here?

ThomasBreuer replied:
@fingolfin The GAP function `HallSubgroup` works also for (finite) nonsolvable groups, and returns either a Hall P-subgroup (if it is unique up to conjugacy), or a list of Hall P-subgroups (representatives of G-conjugacy classes of these subgroups if there is more than one class), or `fail` (if no Hall P-subgroup exists).

For Oscar, we could introduce a function `representatives_hall_subgroups`, which always returns a vector of representatives of G-conjugacy classes of Hall P-subgroups.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -816,16 +1023,20 @@ Return whether `G` is a simple group, i.e.,
 
-Return whether `G` is an almost simple group, i.e., if `S` < `G` < `Aut(S)` for some non-abelian simple group `S`.
+Return whether `G` is an almost simple group,
+i. e., `G` is isomorphic to a Group $H$ with the property
```
```suggestion
i. e., `G` is isomorphic to a group $H$ with the property
```

ThomasBreuer replied:
Yes.

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -842,7 +1053,7 @@ end
     relators(G::FPGroup)
 
-Return a list of relators for the finitely presented group, i.e. elements `[x_1, ... , x_n]` in `F` = `free_group(ngens(G))` such that `G = F/[x_1,...,x_n]`.
+Return an array of relators for the finitely presented group, i.e. elements `[x_1, ... , x_n]` in `F` = `free_group(ngens(G))` such that `G = F/[x_1,...,x_n]`.
```
Wrap long line?

Also, we have just switch most uses of Array to the type aliases Vector and Matrix; so perhaps we should say "a vector" here? Also in a few other places where we mention arrays?

ThomasBreuer replied:
Yes.
(My idea was to replace the GAP style "list" by Julia's "array"; but of course "vector" is more specific.)

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -107,24 +127,28 @@ function abelian_group(v::Vector{Int})
+Here, `T` must be one of `PermGroup`, `FPGroup`, or `PcGroup`.
+
+!!! warning
+    The type need to be specified in the input of the function `abelian_group`, otherwise a group of type `GrpAbFinGen` is returned, which is not a GAP group type. In future versions of Oscar, this may change.
```
Line wrap?

Also, of course we need to properly integrate the abelian groups provided by Hecke.... but that's another PR

ThomasBreuer replied:
Yes.

fingolfin commented on src/Groups/group_constructors.jl:
```diff
@@ -107,24 +127,28 @@ function abelian_group(v::Vector{Int})
 function abelian_group(::Type{T}, v::Vector{Int}) where T <: GAPGroup
-  v1 = GAP.julia_to_gap(v)
-  return T(GAP.Globals.AbelianGroup(_gap_filter(T), v1))
+  return T(GAP.Globals.AbelianGroup(_gap_filter(T), GAP.julia_to_gap(v)))
```
Assuming it works (I didn't test), generally I find this nicer / more "Julia-ish":
```suggestion
  return T(GAP.Globals.AbelianGroup(_gap_filter(T), GAP.GapObj(v)))
```

ThomasBreuer replied:
Yes. (I had already replaced several occurrences.)

fingolfin commented on src/Groups/libraries/primitivegroups.jl:
```diff
@@ -27,7 +29,9 @@ end
-Return the list of all primitive groups (up to isomorphism) satisfying the conditions in `L`. Here, `L` is a vector whose arguments are organized as `L` = [ `func1`, `arg1`, `func2`, `arg2`, ... ], and the function returns all the groups `G` satisfying the conditions `func1`(`G`) = `arg1`, `func2`(`G`) = `arg2`, etc. An argument can be omitted if it corresponds to the boolean value ``true``.
+Return the list of all primitive groups (up to permutation isomorphism)
+satisfying the conditions in `L`.
+Here, `L` is a vector whose arguments are organized as `L` = [ `func1`, `arg1`, `func2`, `arg2`, ... ], and the function returns all the groups `G` satisfying the conditions `func1`(`G`) = `arg1`, `func2`(`G`) = `arg2`, etc. An argument can be omitted if it corresponds to the boolean value `true`.
```
Wrap this line?

ThomasBreuer replied:
Yes.

fingolfin commented on src/Groups/libraries/smallgroups.jl:
```diff
@@ -13,10 +13,10 @@ export
     small_group(n::Int, i::Int)
 
-Return the `i`-th group of order `n` in the catalogue of the GAP Small Groups Library. The group is given of type ``PcGroup`` if the group is solvable, ``PermGroup`` otherwise.
+Return the `i`-th group of order `n` in the catalogue of the GAP Small Groups Library. The group is given of type `PcGroup` if the group is solvable, `PermGroup` otherwise.
```
Wrap?

ThomasBreuer replied:
Yes.

fingolfin commented on src/Oscar.jl:
```diff
@@ -87,6 +87,7 @@ function __init__()
         (GAP.Globals.IsFpGroup, FPGroup),
     ])
     GAP.Packages.load("forms")
+    GAP.Packages.load("ctbllib")
```
What's this about? Is this preparation for moving your character table code into the "official" part of Oscar.jl (you are still working on that, right?)

ThomasBreuer replied:
Yes, this addition accidentally slipped in ...

fingolfin reviewed (commented):
Thanks! Some more minor comments, but this is basically mergeable already.

rfourquet commented on src/Groups/sub.jl:
```diff
@@ -248,21 +254,26 @@ end
 
-Return whether the `G` is solvable.
+Return whether `G` is solvable,
+i. e., whether [`derived_series`](@ref)(`G`)
```
The simpler following version should also work:
```suggestion
i. e., whether [`derived_series(G)`](@ref)
```

ThomasBreuer replied:
@rfourquet Yes, this works, thanks. However, I prefer the "more involved" variant because it creates a hyperlink on "derived_series", whereas your variant creates a hyperlink on "derived_series(G)".

rfourquet commented on src/Groups/GAPGroups.jl:
```diff
@@ -408,9 +606,9 @@ end
+    ngens(G::GAPGroup) -> Int
 
-Return the length of the array gens(G).
+Return the length of the array [`gens`](@ref)`(G)`.
```
```suggestion
Return the length of the array [`gens(G)`](@ref).
```

ThomasBreuer committed:
address @fingolfin's suggestions

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -85,68 +87,191 @@ function parent(x::GAPGroupElem)
+  `G` and its element acts.
+
+!!! note
+    The degree of a group of permutations is not necessarily equal to the largest moved point of the group `G`. For example, the trivial subgroup of `symmetric_group(n)` has degree `n` even though it fixes `n`.
```
```suggestion
    The degree of a group of permutations is not necessarily equal to the
    largest moved point of the group `G`. For example, the trivial subgroup
    of `symmetric_group(n)` has degree `n` even though it fixes `n`.
```

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -816,16 +1112,20 @@ Return whether `G` is a simple group, i.e.,
-Return (``true``,nothing) if `G` is the trivial group, (``true``,``p``) if |`G`| is a non-trivial ``p``-power, (``false``,nothing) otherwise.
+Return `(true, nothing)` if `G` is the trivial group,
+`(true, p)` if |`G`| is a non-trivial power of a prime `p`,
+and `(false, nothing)` otherwise.
```
This is potentially inaccurate, as in current GAP versions, `IsPGroup` may also return `true` for infinite p-groups (not that any are available in GAP, as far as I know...)

fingolfin commented on test/Groups/constructors.jl:
```diff
@@ -128,9 +158,9 @@ end
+               @test index(SO(e,n,q),omega_group(e,n,q))==2
             else
-           #    @test index(GO(e,n,q),omega_group(e,n,q))==2
+               @test index(GO(e,n,q),omega_group(e,n,q))==2
```
Here's a conflict now... please resolve, then merge this

ThomasBreuer committed:
fixed the definition if `ispgroup`

(as suggested by @fingolfin), and rebased

ThomasBreuer commented:
The test with Julia 1.7 on ubuntu ran out of time and got cancelled. I restart the tests.

ThomasBreuer commented:
Now the tests passed (except for the nightly tests), within about one hour.
